### PR TITLE
Get Verified Domains and Skip Shadow Users

### PR DESCRIPTION
**Pull request by cblomart**
Good evening,

I were able to prepare some code that compiles correctly.
Unfortunatly i am not able to test the result as:
- i can't sign the assembly (complains that Nito.AsyncEx is not strong named)
- i suppose the assembly should be in the GAC to be used by sharepoint

Nevertheless i think the logic to filter shadow accounts is sound (naturaly i did it :D).

Would consider merging it in AzureCP?

Cédric

PS: this is one part of the issue #8 i raised.

### CODE REVIEW

**Comment by Yvand** [comment]
Hello,
thank you for your contribution.
Indeed, Nito.AsyncEx is not strongly signed and you need to do it yourself. This is the goal of package [Nivot.StrongNaming](https://www.nuget.org/packages/Nivot.StrongNaming/1.0.4.2) that is also part of the solution.
I took a very quick look at the changes, you seem to duplicate the code just to get tenant details, while I think this could be done in a separate task in QueryAzureADAsync() method, what do you think?

**Comment by cblomart** [comment]
I made separate methods as the return types are not the same.
Plus it did let the original code mostly intact.

I will look at returning a tuple this evening.

**Comment by cblomart** [comment]
Ok, i remplaced the return type of "QueryAzureADAsync" and "QueryAeureADCollectionAsync" to "object". And made the separation of returned types in "BuildFilterAndProcessResultsAsync".

**Comment by cblomart** [comment]
Still not figured out how to compile correctly... maybe still some signing issue...
```
01/26/2018 00:30:32.08 	w3wp.exe (0x0184)                       	0x07A0	SharePoint Foundation         	Claims Authentication         	b280	High    	Error loading the claim provider assembly 'AzureCP, Version=1.0.0.0, Culture=neutral, PublicKeyToken=65dc6b5903b51636' with type 'azurecp.AzureCP'.	c220449e-60ed-204c-7fc8-a7ca75a067a2
01/26/2018 00:30:32.08 	w3wp.exe (0x0184)                       	0x07A0	SharePoint Foundation         	Monitoring                    	b4ly	High    	Leaving Monitored Scope (SPClaimProviderOperations.ClaimsForEntity()). Execution Time=144.232	c220449e-60ed-204c-7fc8-a7ca75a067a2
01/26/2018 00:30:32.08 	w3wp.exe (0x0184)                       	0x07A0	SharePoint Foundation         	Claims Authentication         	af3zp	Unexpected	STS Call Claims Saml: Problem getting output claims identity. Exception: 'System.IO.FileNotFoundException: Could not load file or assembly 'AzureCP, Version=1.0.0.0, Culture=neutral, PublicKeyToken=65dc6b5903b51636' or one of its dependencies. The system cannot find the file specified.  File name: 'AzureCP, Version=1.0.0.0, Culture=neutral, PublicKeyToken=65dc6b5903b51636'     at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)     at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAss...	c220449e-60ed-204c-7fc8-a7ca75a067a2
```

**Comment by Yvand** [comment]
65dc6b5903b51636 is the public key token of standard AzureCP.dll, so SharePoint is telling you that it doesn't find the standard AzureCP.dll.
Your own version of AzureCP.dll has a different public key token because you created your own private key to sign it (and create a strong named assembly).

In short, this error means that something tries to load the AzureCP.dll, so it is very likely that you did not completely uninstall the standard AzureCP.wsp package before you deployed your custom version (which overwrote all feature AzureCP deployed).

This is why I strongly recommended that you completely uninstall standard AzureCP.wsp before you start deploying your custom version.

Now the environment is a little bit messy. To fix it, I would suggest that you try the following;
- Completely retract and remove custom/standard AzureCP.wsp currently deployed
- Visit [this page](http://ldapcp.com/Fix-setup-issues.html) to make sure there is no AzureCP artifact left, and delete them if that's the case
**Note**: this page is for LDAPCP, but the principle is exactly the same for AzureCP, juste replace "LDAPCP" with "AzureCP"
- Once it's cleaned, deploy only your custom version of AzureCP.wsp

HTH

**Comment by cblomart** [comment]
Thanks... Does helps a lot.

Still need to get back the config of all the tenants used to test but i can run the compiled version of AzureCP :-).

I just needed to correct a bug on the casing of the requested properties to filter shadow accounts.
Now people picker does work well with a single tenant. 

Cédric

**Comment by cblomart** [comment]
Any reason you didn't use Nito.AsyncEx.StrongNamed ?

**Comment by Yvand** [comment]
Hello Cédric,

I use Nivot.StrongNaming to sign all unsigned assemblies in Nito.AsyncEx, but they are not added to the project, so you need to sign them too

I just made some cleaning in the code and pushed dev into master. The goal was to publish the release ASAP as it is a lot more reliable than previous one, and it will give me time to review your changes in detail.
Unfortunately it created some conflicts with your PR. I think the easiest is that you sync all changes from Yvand/AzureCP, resolve conflicts and update PR. Can you please do that?
Thanks!

**Comment by cblomart** [comment]
Hello,

I resoled the conficts in github.... I would like to check a rebuild later tonight. But we do have a demo tomorrow morning with Ops & Business so we don't wan't to break things the day before.

"Nito.AsyncEx.StrongNamed" is the same nuget packcage as "Nito.AsyncEx" but with signed assemblies. So you don't have to bother with "Nivot.StrongNaming" (as the nivot.AsyncEx module is signed then).

**Comment by Yvand** [comment]
Ah, indeed, I should have used "Nito.AsyncEx.StrongNamed", I didn't see it unfortunately...
I'll review your changes in detail and give you a feedback, but please be patient, it will take me time

**Comment by Yvand** [comment]
I took a first look at your changes.

Your logic assumes that nobody will ever want to add a guest user if he is in a registered domain. Are you 100% sure about that?

Besides, your logic works for search (e.g. find everyone starting with "user") since it consists of removing results from the list.

But SharePoint may want to validate such guest users (e.g. return 1 PickerEntity for "userX@domain.net"), which won't work anymore. This can't happen with your current implementation, but think about:
- Administrators who already deployed a previous version of AzureCP and added such guest users that you now exclude
- Administrators who may want to add such guest users anyway (back to my initial question)

One way to resolve this is to add a switch to the AzureCP admin page to enable/disable your filtering.
But still I'm very interested if it could make sense, in some scenario, to want to add guest users that are in a registered domain. What do you think?

On a side note, I wonder if we could find a better implementation of AzurecpResult, so that it could properly contain domains too, current solution works but it really looks like a quick & dirty solution.

**Comment by cblomart** [comment]
I have a lenghly answer but let's be a bit shorter:

As per #3 B2B is kinda difficult. Works for microsoft accounts (MSA: google, hotmail) but not for B2B accounts between AADs.
The problem seems here to be in the claims forwarded by AAD.

So with this B2B between AADs you can pick the user from the picker but the access will be given to an "*#ext#@*.onmicrosoft.com" (* is a wildcard) but AAD with send the real mail address as identity.
For MSA it will provider the right claim so when invited you can pick it.

We do want people picker, otherwise just bypassing AAD lookup would have been suffisient.

We do also feel that people trying to use AzureCP with guest users have a hard time having it properly for both MSA and B2B between AADs.

Remember that the filter will only be effective if you have multiple tenant. We can discuss it, but i see no situation where you will invite in AAD a user from your own tenant.

In the end we do think that the filter is a good idea: 
it would prevent to give access that won't work.

Upgrade is a no brainer if you have only one AAD. If you have multiple AAD you might be pleased to 
- remove any modification made on claims in AzureCP and/or ADFS
- have a solution that works for all B2B users (MSA and B2B between AADs)
- have a clean people picker.

Inviting a guest user ("#ext#") from your known AADs? Really?  

I realy hope i can convice you that this filter is nothing bad.

I will nevertheless look at configuration if ok with you:
other than bypassing the filter, being able to indicate which AAD is federated with ADFS is usefull to prevent returning guest users from tenants that are not (the guest account needs to exist in a federated AAD to work)

I may also look further on AzurecpResult:
Maybe one result per tenant search containing a domain list and a directory object list might be better. than a raw object list.

**Comment by Yvand** [comment]
Hi, no worries, I'm already convinced that this is a great improvement in AzureCP.
My main concern is if this should be the default behavior, and configurable (deactivable).
The more I think about it, the more I agree that noone would want to deactivate this, so there is no need to add a switch to the AzureCP admin page.

I made some quick tests and your changes work fine in my environment, but I'll take a much deeper look at the code changes later this week (I hope).
I'll keep you posted, thanks again for your contribution
And yes, using lists in AzureCPResult could be a more elegant solution than current implementation.

**Comment by cblomart** [comment]
I just added a switch to tell wich tenant(s) are federated with ADFS. For these tenants Guests will be ignored.

**Comment by Yvand** [comment]
Hi Cédric,
a guest user won't be removed if his original domain (retrieved from his email) is not found in another AAD tenant registered in AzureCP.
So only AAD Guest users that have their email suffix matching another AAD tenant registered in AzureCP are removed from results.
My main concern was if there would be any reason to not remove them, and you convinced me that no.

But in your last commit you added an option to activate/deactivate this filtering.
Does that mean you think otherwise now?

Note: I found [this article about B2B users](https://docs.microsoft.com/en-us/azure/active-directory/active-directory-b2b-user-properties), it helps me to better understand the differences between both Guests and Members.

**Comment by cblomart** [comment]
You are right in your first sentences.

Notice the last commit doesn't remove the filter.

What is does is indicate to only search for member user or all users for each tenant.

**Why doing that?**

The main scenario we are looking at is as follows:

```asciiart
                                           +----------------+
   +........AzureCP AAD Graph API..........| AAD seconday 1 | 
   :                                :     /+----------------+ 
+----+ SAML 1.0 +------+ SAML 2.0 +-------------+     +----------------+
| SP | -------- | ADFS | -------- | AAD primary |-B2B-| Guest Accounts |
+----+          +------+          +-------------+     +----------------+
   :                                      \+-----------------+
   +........AzureCP AAD Graph API..........| AAD secondary 2 |
                                           +-----------------+
```

What i know:
Guest can be in any tenants.
AAD can only authentify users he knows, is this a Member or a Guest.

If guest from a secondary tenant tries to login to sharepoint (s)he will be redirected to the primary AAD. 
This AAD will complain that there is no account for the guest user. 
Even if invited on both tenants, the guest will have a different UPN in both (guest1_gmail.com#ext#@Alice.onMicrosoft.com and guest1_gmail.com#ext#@Bob.onMicrosoft.com)

Federating to multiple AADs is possible. 

> REM: 
>
> This is some work to register sharepoint in each AAD.
> This will get confusing for guest users at home real discovery.
> The only advantage i see is lessen the invitation work

```asciiart
   ,....AAD Graph API..,
   :                   :
   :                +-------+   +----------------+
   :                | AAD 1 | - | Guest Accounts |
+----+    +------+ /+-------+   +----------------+
| SP | -- | ADFS |     | (optional B2B)
+----+    +------+ \+-------+   +----------------+
   :                | AAD 1 | - | Guest Accounts |
   :                +-------+   +----------------+
   :                   :
   '....AAD Graph API..'
```

So what i conclude:
**Guest users should only be used on AAD federated with Sharepoint**

Indicating with wich AAD a Sharepoint (trought ADFS) is federated allows to filter out guest users that should not be given access.

**Comment by Yvand** [comment]
Ok, then let's give it a name more accurate to what it's actually doing:
Can you call the property "MembersUserTypeOnly"?

**Comment by cblomart** [comment]
Hello here what should do the trick.
Thansk for keeping up!

**Comment by Yvand** [comment]
When I visit AzureCP settings page, I get this error:
```
A field or property with the name 'MemberUserTypeOnly' was not found on the selected data source. 

[HttpException (0x80004005): A field or property with the name 'MemberUserTypeOnly' was not found on the selected data source.]
System.Web.UI.WebControls.BoundField.TryGetSimplePropertyValue(Object dataItem, Object& data) +373
System.Web.UI.WebControls.BoundField.GetValue(Control controlContainer) +261
System.Web.UI.WebControls.BoundField.OnDataBindField(Object sender, EventArgs e) +71
System.Web.UI.Control.OnDataBinding(EventArgs e) +83
System.Web.UI.Control.DataBind(Boolean raiseOnDataBinding) +167
System.Web.UI.Control.DataBindChildren() +249
System.Web.UI.Control.DataBind(Boolean raiseOnDataBinding) +184
System.Web.UI.WebControls.GridView.CreateRow(Int32 rowIndex, Int32 dataSourceIndex, DataControlRowType rowType, DataControlRowState rowState, Boolean dataBind, Object dataItem, DataControlField[] fields, TableRowCollection rows, PagedDataSource pagedDataSource) +247
System.Web.UI.WebControls.GridView.CreateChildControls(IEnumerable dataSource, Boolean dataBinding) +5138
Microsoft.SharePoint.WebControls.SPGridView.CreateChildControls(IEnumerable dataSource, Boolean dataBinding) +92
System.Web.UI.WebControls.CompositeDataBoundControl.PerformDataBinding(IEnumerable data) +97
System.Web.UI.WebControls.GridView.PerformDataBinding(IEnumerable data) +18
System.Web.UI.WebControls.DataBoundControl.OnDataSourceViewSelectCallback(IEnumerable data) +198
azurecp.PropertyCollectionBinder.BindGrid(SPGridView grid) +114
azurecp.AzureCPSettings.PopulateLdapConnectionGrid() +408
azurecp.AzureCPSettings.PopulateFields() +75
azurecp.AzureCPSettings.Page_Load(Object sender, EventArgs e) +1882
```

It seems to be the same issue as [described here](https://stackoverflow.com/questions/17120505/error-a-field-or-property-with-the-name-was-not-found-on-the-selected-data-sou), do you repro this error?

**Comment by cblomart** [comment]
Will check this evening.

2018-02-12 10:02 GMT+01:00 Yvan Duhamel <notifications@github.com>:

File: AzureCP/AzureCP.cs [hunk 1]
```diff
@@ -491,24 +491,83 @@ public void BuildFilterAndProcessResultsAsync(string input, List<AzureADObject>
...
+                    //skip shadow users (type is guest and mail domain in tenants verified domains)
+                    string userType = GetGraphPropertyValue(searchResult.DirectoryObjectResult, "UserType");
+                    if (String.IsNullOrEmpty(userType))
+                    {
+                        AzureCPLogging.Log(
+                            String.Format("[{0}] BuildFilterAndProcess skipped user because usertype is empty.", ProviderInternalName),
+                            TraceSeverity.Unexpected, EventSeverity.Warning, AzureCPLogging.Categories.Lookup);
+                        continue;
+                    }
+                    if (String.Equals(userType, "Guest", StringComparison.InvariantCultureIgnoreCase))
```
Comment (Yvand) [inline AzureCP/AzureCP.cs]: Can you make "Guest" a string in azurecp.Constants class?

File: AzureCP/AzureCP.cs [hunk 2]
```diff
@@ -705,32 +767,31 @@ private ActiveDirectoryClient RefreshAzureADContext(AzureTenant coco)
...
+                                // get the user search
+                                IReadOnlyQueryableSet<IUser> userQuerySet =  coco.ADClient.Users.Where(userQuery);
+                                // if need filter out guest users
+                                if (coco.MemberUserTypeOnly) 
+                                {
+                                    AzureCPLogging.Log(String.Format("[{0}] UserQueryTask filtering members only for tenant '{1}'", ProviderInternalName, coco.TenantName), 
+                                        TraceSeverity.Medium, EventSeverity.Information, AzureCPLogging.Categories.Lookup);
+                                    // "where not" not supported in OData...
+                                    //userQuerySet = userQuerySet.Where(x => !x.UserType.Equals("Guest",StringComparison.InvariantCultureIgnoreCase));
+                                    userQuerySet = userQuerySet.Where(x => x.UserType.Equals("Member",StringComparison.InvariantCultureIgnoreCase));
```
Comment (Yvand) [inline AzureCP/AzureCP.cs]: Can you make "Member" a string in azurecp.Constants class?

File: AzureCP/TEMPLATE/ADMIN/AzureCP/AzureCPSettings.aspx.cs [hunk 3]
```diff
@@ -414,14 +415,16 @@ public PropertyCollectionBinder()
             PropertyCollection.Columns.Add("Id", typeof(Guid));
             PropertyCollection.Columns.Add("TenantName", typeof(string));
             PropertyCollection.Columns.Add("ClientID", typeof(string));
+            PropertyCollection.Columns.Add("MemberUserTypeOnly", typeof(bool));
         }
 
-        public void AddRow(Guid Id, string TenantName, string ClientID)
+        public void AddRow(Guid Id, string TenantName, string ClientID, bool Federated)
```
Comment (Yvand) [inline AzureCP/TEMPLATE/ADMIN/AzureCP/AzureCPSettings.aspx.cs]: Can you rename Federated into MemberUserTypeOnly?

File: AzureCP/AzureCP.cs [hunk 4]
```diff
@@ -491,24 +491,83 @@ public void BuildFilterAndProcessResultsAsync(string input, List<AzureADObject>
...
+            // search users, groups and domains
+            Task<List<object>> searchResultsTask = this.QueryAzureADCollectionAsync(input, userQuery, groupQuery);
             searchResultsTask.ConfigureAwait(false);
+            // wait for task
             searchResultsTask.Wait();
-            List<AzurecpResult> searchResults = searchResultsTask.Result;
-            if (searchResults == null || searchResults.Count == 0) return;
+            // Separate AzurecpResult from string (domains)
+            IEnumerable<AzurecpResult> searchResults = searchResultsTask.Result.OfType<AzurecpResult>();
+            IEnumerable<string> domains = searchResultsTask.Result.OfType<string>();
```
Comment (Yvand) [inline AzureCP/AzureCP.cs]: Current solution works great, but as discussed earlier, it's really not elegant. 
I like your suggestion to split domains into a separate list. Do you plan to do it?

**Comment by cblomart** [comment]
spot on!
I'll look into it.
cosmetic changes will be done rapidly.
Remodling the Azure AD queries will take a bit of time.

**Comment by cblomart** [comment]
Here, this should answer the different remarks.

**Pull request merged**